### PR TITLE
Refactoring of Middleware for ASP.NET Core provider

### PR DESCRIPTION
**Pull request by phillip-haydon**
### v10.0.0
- RaygunClient for NET Core can now be treated as a Singleton
- Changed the Middleware for AspNetCore
  - See: https://github.com/MindscapeHQ/raygun4net/pull/518 
  - Updated the registration of `.AddRaygun()`
  - Updated the usage of `.UseRaygun()`
  - Removed `RaygunMiddlewareSettings` and `RaygunClientProvider` in favour of `IRaygunUserProvider`
  - Removed custom code for maintaining request body in the middleware and used `Request.EnableBuffering()` instead
  - Deprecated multiple settings in RaygunClient in favour of using `RaygunSettings`
  - Deprecated `UserInfo` property in favour of `IRaygunUserProvider`
  - Deprecated `RaygunMiddlewareOptions` in favour of `RaygunSettings`
  - Introduced new `.AddRaygunUserProvider()` to register a default implementation
  - Introduced new `.AddRaygunUserProvider<T>()` to register a custom implementation
- Fixed bug with filters where uppercase properties were compared against lowercase filters

# Changes

The primary change is that RaygunClient can now treated as a singleton, this changed required some shuffling of things and making some things obsolete to remove duplication and consolidation.

## Updates to `AddRaygun(...)`

The old implementation assumed that all configuration was done in the appsettings.json and allowed for no code-overrides that were easily done. 

```csharp
builder.Services.AddRaygun(builder.Configuration, new RaygunMiddlewareSettings
{
  // adds an optional example of over riding the client provider
  ClientProvider = new ExampleRaygunAspNetCoreClientProvider()
});
```

**_Breaking Change_** - when using RaygunMiddlewareSettings

The new API removes RaygunMiddlewareSettings argument and allows for:

```csharp
// Configure using appsettings.json
builder.Services.AddRaygun(builder.Configuration);

// or

// Configure using appsettings.json -> and supplement / override with supplied settings
builder.Services.AddRaygun(builder.Configuration, settings =>
{
  settings....
});

// or

// Configure with just supplied settings
builder.Services.AddRaygun(settings =>
{
  settings....
});
```

## Updates to ClientProvider

The middleware took in a custom client provider, the primary use-case for this was to include the current user and assign this to `UserInfo` property on the RaygunClient, if the client is treated as a singleton then this doesn't work. 

We have introduced a new interface `IRaygunUserProvider`, and ASP .NET Core has a default implementation which can be opted into. 

```csharp
builder.Services.AddRaygunUserProvider();
```

This will register `DefaultRaygunUserProvider` which will capture the `name` and `email` from the claims. If you want you can create your own implementation and register it.

```csharp
builder.Services.AddRaygunUserProvider<CustomUserProvider>();
```

**_Breaking Change_** - when using RaygunClient as a singleton in non-web applications (desktop app)

Prior to this change you could do:

```csharp
client.UserInfo = new RaygunIdentifierMessage(...);
```

The work-around is to create an implementation of IRaygunUserProvider and assign it in the ctor, or register it in the DI container so it can be resolved when the client is created.

```csharp
var client = new RaygunClient(settings, new CustomUserProvider());
```

### CODE REVIEW

File: Mindscape.Raygun4Net.NetCore.Common/RaygunClientBase.cs [hunk 1]
```diff
@@ -367,24 +367,23 @@ public Task SendInBackground(RaygunMessage raygunMessage)
...
-        .Build();
+                                        .Customise(customiseMessage)
+                                        .SetEnvironmentDetails()
+                                        .SetMachineName(Environment.MachineName)
+                                        .SetExceptionDetails(exception)
+                                        .SetClientDetails()
+                                        .SetVersion(_settings.ApplicationVersion)
+                                        .SetTags(tags)
+                                        .SetUserCustomData(userCustomData)
+                                        .SetUser(userInfo ?? _userProvider?.GetUser())
```
[the remark this answers is not in the document]
Comment (phillip-haydon) [reply Mindscape.Raygun4Net.NetCore.Common/RaygunClientBase.cs:384]: userInfo here is the argument not the property.

File: Mindscape.Raygun4Net.AspNetCore/README.md [hunk 2]
```diff
@@ -117,49 +75,93 @@ public class RaygunController : Controller
...
     {
-      _singletonRaygunClient.SetCurrentContext(HttpContext);
-      await _singletonRaygunClient.Send(ex);
+      return null;
     }
 
-    return View();
+    var identity = ctx.User.Identity as ClaimsIdentity;
+    
+    if (identity?.IsAuthenticated == true)
```
Comment (xenolightning) [inline Mindscape.Raygun4Net.AspNetCore/README.md:117]: Could maybe be a pattern expression if you want to be slightly sadistic

File: Mindscape.Raygun4Net.Mvc/Mindscape.Raygun4Net.Mvc.csproj
[the remark this answers is not in the document]
Comment (phillip-haydon) [reply Mindscape.Raygun4Net.Mvc/Mindscape.Raygun4Net.Mvc.csproj:1]: I was trying to remove the security warning. And the license location was not consistent.

File: Mindscape.Raygun4Net.NetCore.Common/RaygunClientBase.cs [hunk 3]
```diff
@@ -52,71 +52,64 @@ public abstract class RaygunClientBase
...
     /// Gets or sets the user identity string.
     /// </summary>
+    [Obsolete("Provide a `IRaygunUserProvider` to the RaygunClient constructor instead.")]
     public virtual string User { get; set; }
 
     /// <summary>
     /// Gets or sets information about the user including the identity string.
     /// </summary>
+    [Obsolete("Provide a `IRaygunUserProvider` to the RaygunClient constructor instead.")]
     public virtual RaygunIdentifierMessage UserInfo { get; set; }
```
Comment (xenolightning) [inline Mindscape.Raygun4Net.NetCore.Common/RaygunClientBase.cs:62]: We should probably still use this in the message builder if it's set

File: Mindscape.Raygun4Net.NetCore.Common/RaygunClientBase.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (xenolightning) [reply Mindscape.Raygun4Net.NetCore.Common/RaygunClientBase.cs:384]: `info ?? provider ?? info ?? new Info(user)`

🤘🏼

File: Raygun4Net.AspNetCore.Tests/RaygunMiddlewareTests.cs [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+      Headers =
+      {
+        { "Banana", "yellow" },
+        { "Apple", "red" }
+      }
+    });
+
+    await act.Should().ThrowAsync<Exception>().WithMessage("Banana's are indeed yellow");
+
+    await Task.Delay(500);
```
Comment (xenolightning) [inline Raygun4Net.AspNetCore.Tests/RaygunMiddlewareTests.cs:164]: One day we'll get rid of delays in test 😺

**Review by xenolightning — DISMISSED** [review]
Looking pretty good.

I think we still support the existing user info mechanism until it's fully removed

File: Mindscape.Raygun4Net.AspNetCore/ApplicationBuilderExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    services.TryAddSingleton(s => new RaygunClient(s.GetService<RaygunSettings>()!, s.GetService<IRaygunUserProvider>()));
+    services.AddHttpContextAccessor();
+
+    return services;
+  }
+
+  /// <summary>
+  /// Registers the Raygun Client and Raygun Settings with the DI container. Settings will be defaulted and overridden by providing a custom configuration delegate.
+  /// </summary>
+  public static IServiceCollection AddRaygun(this IServiceCollection services, Action<RaygunSettings>? options)
```
Comment (PanosNB) [inline Mindscape.Raygun4Net.AspNetCore/ApplicationBuilderExtensions.cs:63]: Shouldn't this `AddRaygun` be calling the one above after creating and passing an `ICofiguration` to minimize code duplication?

File: Mindscape.Raygun4Net.AspNetCore/Builders/RaygunAspNetCoreRequestMessageBuilder.cs [hunk 6]
```diff
@@ -334,5 +345,52 @@ private static IDictionary ToDictionary(IFormCollection query, Func<string, bool
...
+
+    internal static bool IsIgnored(string? key, IReadOnlyList<string> list)
+    {
+      if (key == null || (list.Count == 1 && "*".Equals(list[0])))
+      {
+        return true;
+      }
+
+      // ReSharper disable once LoopCanBeConvertedToQuery
+      foreach (var ignoredKey in list)
```
Comment (PanosNB) [inline Mindscape.Raygun4Net.AspNetCore/Builders/RaygunAspNetCoreRequestMessageBuilder.cs:357]: What exactly are we ignoring here? Is the linear search a concern: do we need a dictionary lookup or another faster algorithm?

File: Mindscape.Raygun4Net.AspNetCore/DefaultRaygunUserProvider.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    var ctx = _contextAccessor.HttpContext;
+    
+    if (ctx == null)
+    {
+      return null;
+    }
+
+    var identity = ctx.User.Identity as ClaimsIdentity;
+    
+    if (identity?.IsAuthenticated == true)
```
Comment (PanosNB) [inline Mindscape.Raygun4Net.AspNetCore/DefaultRaygunUserProvider.cs:28]: I prefer not writing equals true or false for Boolean checks but whatever is in our style guide

File: Mindscape.Raygun4Net.NetCore.Common/Messages/RaygunIdentifierMessage.cs [hunk 8]
```diff
@@ -2,11 +2,18 @@
 {
   public class RaygunIdentifierMessage
   {
+#if DEBUG
```
Comment (PanosNB) [inline Mindscape.Raygun4Net.NetCore.Common/Messages/RaygunIdentifierMessage.cs:5]: Ensure this is documented in our public docs

File: Mindscape.Raygun4Net.NetCore.Tests/RaygunMessageBuilderTests.cs [hunk 9]
```diff
@@ -104,5 +103,24 @@ public void ResponseIsNullForNonWebExceptions()
       RaygunMessage message = _builder.Build();
       Assert.That(message.Details.Response, Is.Null);
     }
+    
+    [Test]
+    public void Customise_ExistingMessage_CorrectlyModifiesProperties()
```
Comment (PanosNB) [inline Mindscape.Raygun4Net.NetCore.Tests/RaygunMessageBuilderTests.cs:108]: I won't make you change it but please stick to American English in the future such that our code is homogeneous :-)

File: Raygun4Net.AspNetCore.Tests/RaygunAspNetCoreRequestMessageBuilderTests.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    RaygunAspNetCoreRequestMessageBuilder.IsIgnored(key, new[] { ignoredKey }).Should().Be(expected);
+  }
+  
+  [TestCase("AppLe", "apple", true)]
+  [TestCase("APPLE", "apple", true)]
+  [TestCase("apple", "aPPle", true)]
+  [TestCase("apple", "APPLE", true)]
+  [TestCase("apple", "apple", true)]
+  [TestCase("APPLE", "APPLE", true)]
+  [TestCase("apple", "banana", false)]
```
Comment (PanosNB) [inline Raygun4Net.AspNetCore.Tests/RaygunAspNetCoreRequestMessageBuilderTests.cs:35]: If only my son Zakros could see this! Hahaha

**Review by PanosNB — APPROVED** [review]
All good on my end. Just some minor comments/questions for you to consider. Great work!

**Pull request merged**